rust/crd/src/lib.rs

```diff
@@ -46,12 +59,26 @@ use stackable_operator::config::fragment;
 pub struct HdfsClusterSpec {
     pub image: ProductImage,
     pub auto_format_fs: Option<bool>,
-    pub zookeeper_config_map_name: String,
-    pub data_nodes: Option<Role<DataNodeConfig>>,
-    pub name_nodes: Option<Role<NameNodeConfig>>,
-    pub journal_nodes: Option<Role<JournalNodeConfig>>,
     pub dfs_replication: Option<u8>,
-    pub log4j: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub name_nodes: Option<Role<NameNodeConfigFragment>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub data_nodes: Option<Role<DataNodeConfigFragment>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub journal_nodes: Option<Role<JournalNodeConfigFragment>>,
+    /// Name of the Vector aggregator discovery ConfigMap.
+    /// It must contain the key `ADDRESS` with the address of the Vector aggregator.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub vector_aggregator_config_map_name: Option<String>,
+    /// Name of the ZooKeeper discovery config map.
+    pub zookeeper_config_map_name: String,
+}
+
+/// This is a shared trait for all role/role-group config structs to avoid duplication
+/// when extracting role specific configuration structs.
+pub trait MergedConfig {
+    fn resources(&self) -> Resources<HdfsStorageConfig, NoRuntimeLimits>;
+    fn logging(&self) -> Logging<Container>;
 }
```

Comment:
Just wanted to mention that we need different resources for Datanodes (regarding pvcs) in https://github.com/stackabletech/hdfs-operator/pull/291

Reply:
Yeah it looks like a lot of overhead here for now since all three configs are the same. 
Normally i would assume that different roles take different configurations. We could stick all of this into one config, but e.g. creating docs out of our CRD will document settings for role_1 that are not possible for role_2 etc.

For HDFS until the datanode pvcs that @sbernauer mentioned come into play (there is a PR up already) it is overkill, but i wanted to get a first implementation for different configs in different roles out.